File: conformance/harness/ruby/vbp_harness.rb
#!/usr/bin/env ruby
# frozen_string_literal: true
#
# vbp_harness.rb — VBP v1 conformance skeleton harness (Ruby).
#
# Loads conformance/vbp_suite.yaml using only the Ruby stdlib
# (a hand-rolled block-style YAML parser; we deliberately do NOT
# use the 'psych' gem, although Ruby ships with it, because we
# want the harness to be runnable in any environment with a bare
# `ruby` interpreter). Iterates every test, emits a JUnit XML
# report, and SKIPs all tests. Exit code 0 on success, 1 on any
# FAIL/ERROR.
#
# Usage:
#   ruby vbp_harness.rb --suite ../../vbp_suite.yaml \
#                       --out   ./vbp-conformance-ruby.junit.xml

require 'optparse'
require 'fileutils'
require 'time'

# ---------------------------------------------------------------------------
# Tiny block-style YAML loader (stdlib-only).
# Mirrors the structure used by the other 7 harnesses' loaders.
# ---------------------------------------------------------------------------

def parse_scalar(v)
  v = v.to_s.strip
  return nil if v.empty?
  low = v.downcase
  return true  if low == 'true'
  return false if low == 'false'
  return nil   if low == 'null' || low == '~'
  return v[1..-2] if (v.start_with?('"') && v.end_with?('"')) ||
                     (v.start_with?("'") && v.end_with?("'"))
  if v.start_with?('[') && v.end_with?(']')
    inner = v[1..-2].strip
    return [] if inner.empty?
    return inner.split(',').map { |p| parse_scalar(p.strip) }
  end
  return Integer(v) if v =~ /\A-?\d+\z/
  return Float(v)   if v =~ /\A-?\d+\.\d+\z/
  v
end

def leading_spaces(s)
  s.length - s.lstrip.length
end

def load_yaml(path)
  lines = File.readlines(path)
  out = {}
  tests = []
  i = 0
  while i < lines.length
    line = lines[i]
    s = line.strip
    if s.empty? || s.start_with?('#')
      i += 1
      next
    end
    indent = leading_spaces(line)
    if s.start_with?('- ') && indent == 2
      cur = {}
      first = s[2..].strip
      if first.include?(':')
        k, v = first.split(':', 2)
        cur[k.strip] = parse_scalar(v)
      end
      i += 1
      while i < lines.length
        nx = lines[i]
        if nx.strip.empty?
          i += 1
          next
        end
        stripped = nx.lstrip
        ix = nx.length - stripped.length
        break if ix == 2 && stripped.start_with?('- ')
        break if ix == 0 && !stripped.empty?
        if stripped.include?(':')
          k, v = stripped.split(':', 2)
          cur[k.strip] = parse_scalar(v)
        end
        i += 1
      end
      tests << cur
      next
    end
    if indent == 0 && s.include?(':')
      k, v = s.split(':', 2)
      out[k.strip] = parse_scalar(v)
    end
    i += 1
  end
  out['tests'] = tests unless tests.empty?
  out
end

# ---------------------------------------------------------------------------
# JUnit emit
# ---------------------------------------------------------------------------

def xml_escape(s)
  s.to_s
   .gsub('&', '&amp;')
   .gsub('<', '&lt;')
   .gsub('>', '&gt;')
   .gsub('"', '&quot;')
   .gsub("'", '&apos;')
end

def write_junit(outcomes, out_path, suite_name)
  by_cat = outcomes.group_by { |o| o[:category] }
  File.open(out_path, 'w') do |f|
    f.puts '<?xml version="1.0" encoding="UTF-8"?>'
    f.puts '<testsuites>'
    by_cat.keys.sort.each do |cat|
      oo = by_cat[cat]
      fails = oo.count { |o| o[:status] == 'fail' }
      skips = oo.count { |o| o[:status] == 'skip' }
      errs  = oo.count { |o| o[:status] == 'error' }
      total_dur = oo.sum { |o| o[:duration] }
      f.puts "  <testsuite name=\"#{xml_escape(cat)}\" tests=\"#{oo.length}\" failures=\"#{fails}\" skipped=\"#{skips}\" errors=\"#{errs}\" time=\"#{format('%.3f', total_dur)}\">"
      oo.each do |o|
        f.puts "    <testcase classname=\"#{xml_escape(suite_name)}\" name=\"#{xml_escape("#{o[:id]} #{o[:name]}")}\" time=\"#{format('%.3f', o[:duration])}\">"
        f.puts "      <failure>#{xml_escape(o[:message])}</failure>"  if o[:status] == 'fail'
        f.puts "      <skipped>#{xml_escape(o[:message])}</skipped>"  if o[:status] == 'skip'
        f.puts "      <error>#{xml_escape(o[:message])}</error>"      if o[:status] == 'error'
        f.puts '    </testcase>'
      end
      f.puts '  </testsuite>'
    end
    f.puts '</testsuites>'
  end
end

# ---------------------------------------------------------------------------
# Test runner (skeleton — all SKIP)
# ---------------------------------------------------------------------------

def run_test(t)
  {
    id:       t['id'] || 0,
    name:     t['name'] || 'unknown',
    category: t['category'] || 'unknown',
    status:   'skip',
    message:  'Ruby harness: skeleton — no test cases driven end-to-end (TODO: port vbp_harness.go to ruby)',
    duration: 0.0,
  }
end

# ---------------------------------------------------------------------------
# CLI
# ---------------------------------------------------------------------------

opts = {
  suite: 'conformance/vbp_suite.yaml',
  addr:  '127.0.0.1:6380',
  out:   'vbp-conformance-ruby.junit.xml',
  user:  'admin',
  pass:  'TestPassword123!',
  cat:   '',
}
OptionParser.new do |o|
  o.on('--suite S')    { |v| opts[:suite] = v }
  o.on('--addr A')     { |v| opts[:addr] = v }
  o.on('--out F')      { |v| opts[:out] = v }
  o.on('--user U')     { |v| opts[:user] = v }
  o.on('--pass P')     { |v| opts[:pass] = v }
  o.on('--category C') { |v| opts[:cat] = v }
end.parse!

unless File.exist?(opts[:suite])
  warn "ERROR: suite file not found: #{opts[:suite]}"
  exit 2
end

data = load_yaml(opts[:suite])
tests = data['tests'] || []
tests = tests.select { |t| t['category'] == opts[:cat] } unless opts[:cat].empty?
suite_name = data['suite'] || 'vbp-conformance-v1'

outcomes = tests.map { |t| run_test(t) }
write_junit(outcomes, opts[:out], suite_name)

pass_n = outcomes.count { |o| o[:status] == 'pass' }
fail_n = outcomes.count { |o| o[:status] == 'fail' }
skip_n = outcomes.count { |o| o[:status] == 'skip' }
err_n  = outcomes.count { |o| o[:status] == 'error' }
puts 'VBP v1 conformance (Ruby skeleton)'
puts "  tests:  #{outcomes.length}"
puts "  pass:   #{pass_n}"
puts "  fail:   #{fail_n}"
puts "  skip:   #{skip_n}"
puts "  error:  #{err_n}"
puts "  report: #{opts[:out]}"
exit(fail_n + err_n > 0 ? 1 : 0)

File: conformance/harness/node/vbp_harness.js
#!/usr/bin/env node
// vbp_harness.js — VBP v1 conformance skeleton harness (Node.js).
//
// Loads conformance/vbp_suite.yaml using only stdlib (Node 18+ has
// no built-in YAML parser, so this skeleton uses a small block-style
// parser that handles the structure of vbp_suite.yaml).
//
// Iterates every test, emits a JUnit XML report, and SKIPs all
// tests. Exit code 0 on success, 1 on any FAIL/ERROR.
//
// Usage:
//
//   node vbp_harness.js \
//     --suite ../../vbp_suite.yaml \
//     --addr  127.0.0.1:6380 \
//     --out   ./vbp-conformance-node.junit.xml

'use strict';

const fs = require('fs');
const path = require('path');

// ---------------------------------------------------------------------------
// Tiny block-style YAML loader (stdlib-only).
// Mirrors the structure of the python skeleton's _flat_yaml_load —
// both languages share the constraint of "no extra deps".
// ---------------------------------------------------------------------------

function parseScalar(v) {
  if (v === '' || v == null) return null;
  const low = String(v).toLowerCase();
  if (low === 'true') return true;
  if (low === 'false') return false;
  if (low === 'null' || low === '~') return null;
  if (v.startsWith('"') && v.endsWith('"')) return v.slice(1, -1);
  if (v.startsWith("'") && v.endsWith("'")) return v.slice(1, -1);
  if (v.startsWith('[') && v.endsWith(']')) {
    const inner = v.slice(1, -1).trim();
    if (!inner) return [];
    return inner.split(',').map(p => parseScalar(p.trim()));
  }
  if (v.startsWith('{') && v.endsWith('}')) {
    const inner = v.slice(1, -1).trim();
    if (!inner) return {};
    const out = {};
    for (const pair of inner.split(',')) {
      const ix = pair.indexOf(':');
      if (ix < 0) continue;
      out[pair.slice(0, ix).trim()] = parseScalar(pair.slice(ix + 1).trim());
    }
    return out;
  }
  const asInt = Number(v);
  if (Number.isInteger(asInt)) return asInt;
  const asFloat = Number(v);
  if (!Number.isNaN(asFloat)) return asFloat;
  return v;
}

function loadYaml(file) {
  const lines = fs.readFileSync(file, 'utf-8').split(/\r?\n/);
  const out = {};
  const tests = [];
  let i = 0;
  while (i < lines.length) {
    let line = lines[i];
    const s = line.trim();
    if (!s || s.startsWith('#')) { i++; continue; }
    // Compute the indent of the *non-blank* current line. We only
    // treat a `- ` as a top-level test marker when its indent is 2
    // (the suite uses 2-space indent under `tests:`).
    const indent = line.length - line.replace(/^\s+/, '').length;
    if (s.startsWith('- ') && indent === 2) {
      const cur = {};
      const first = s.slice(2).trim();
      if (first.includes(':')) {
        const [k, ...rest] = first.split(':');
        cur[k.trim()] = parseScalar(rest.join(':').trim());
      }
      i++;
      while (i < lines.length) {
        const nxt = lines[i];
        if (!nxt.trim()) { i++; continue; }
        const stripped = nxt.replace(/^\s+/, '');
        const ix = nxt.length - stripped.length;
        // A same-level (indent==2) `- ` starts a new test.
        if (ix === 2 && stripped.startsWith('- ')) break;
        // An indent-0 (or empty) line ends the test block.
        if (ix === 0 && stripped) break;
        if (stripped.includes(':')) {
          const [k, ...rest] = stripped.split(':');
          cur[k.trim()] = parseScalar(rest.join(':').trim());
        }
        i++;
      }
      tests.push(cur);
      continue;
    }
    if (s.includes(':') && indent === 0) {
      const [k, ...rest] = s.split(':');
      out[k.trim()] = parseScalar(rest.join(':').trim());
    }
    i++;
  }
  if (tests.length) out.tests = tests;
  return out;
}

// ---------------------------------------------------------------------------
// JUnit emit
// ---------------------------------------------------------------------------

function xmlEscape(s) {
  return String(s)
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&apos;');
}

function writeJUnit(outcomes, outPath, suiteName) {
  const byCat = {};
  for (const o of outcomes) {
    (byCat[o.category] = byCat[o.category] || []).push(o);
  }
  const lines = ['<?xml version="1.0" encoding="UTF-8"?>', '<testsuites>'];
  for (const cat of Object.keys(byCat).sort()) {
    const oo = byCat[cat];
    const totalDur = oo.reduce((a, b) => a + b.duration, 0);
    const fails = oo.filter(x => x.status === 'fail').length;
    const skips = oo.filter(x => x.status === 'skip').length;
    const errs  = oo.filter(x => x.status === 'error').length;
    lines.push(`  <testsuite name="${xmlEscape(cat)}" tests="${oo.length}" failures="${fails}" skipped="${skips}" errors="${errs}" time="${totalDur.toFixed(3)}">`);
    for (const o of oo) {
      lines.push(`    <testcase classname="${xmlEscape(suiteName)}" name="${xmlEscape(o.id + ' ' + o.name)}" time="${o.duration.toFixed(3)}">`);
      if (o.status === 'fail') {
        lines.push(`      <failure>${xmlEscape(o.message)}</failure>`);
      } else if (o.status === 'skip') {
        lines.push(`      <skipped>${xmlEscape(o.message)}</skipped>`);
      } else if (o.status === 'error') {
        lines.push(`      <error>${xmlEscape(o.message)}</error>`);
      }
      lines.push(`    </testcase>`);
    }
    lines.push(`  </testsuite>`);
  }
  lines.push('</testsuites>');
  fs.writeFileSync(outPath, lines.join('\n') + '\n');
}

// ---------------------------------------------------------------------------
// Test runner (skeleton — all SKIP)
// ---------------------------------------------------------------------------

function runTest(t /*, addr, user, pass */) {
  return {
    id: t.id || 0,
    name: t.name || 'unknown',
    category: t.category || 'unknown',
    status: 'skip',
    message: 'Node harness: skeleton — no test cases driven end-to-end (TODO: port vbp_harness.go to node)',
    duration: 0,
  };
}

// ---------------------------------------------------------------------------
// CLI
// ---------------------------------------------------------------------------

function main() {
  const args = process.argv.slice(2);
  const opts = {
    suite: 'conformance/vbp_suite.yaml',
    addr: '127.0.0.1:6380',
    out: 'vbp-conformance-node.junit.xml',
    user: 'admin',
    pass: 'TestPassword123!',
    category: '',
  };
  for (let i = 0; i < args.length; i++) {
    const k = args[i];
    const v = args[i + 1];
    switch (k) {
      case '--suite':    opts.suite = v; i++; break;
      case '--addr':     opts.addr = v; i++; break;
      case '--out':      opts.out = v; i++; break;
      case '--user':     opts.user = v; i++; break;
      case '--pass':     opts.pass = v; i++; break;
      case '--category': opts.category = v; i++; break;
    }
  }

  if (!fs.existsSync(opts.suite)) {
    console.error(`ERROR: suite file not found: ${opts.suite}`);
    process.exit(2);
  }

  const suite = loadYaml(opts.suite);
  let tests = suite.tests || [];
  if (opts.category) tests = tests.filter(t => t.category === opts.category);
  const suiteName = suite.suite || 'vbp-conformance-v1';

  const outcomes = tests.map(t => runTest(t, opts.addr, opts.user, opts.pass));

  try {
    writeJUnit(outcomes, opts.out, suiteName);
  } catch (e) {
    console.error(`ERROR: write JUnit: ${e.message}`);
    process.exit(2);
  }

  const passN = outcomes.filter(o => o.status === 'pass').length;
  const failN = outcomes.filter(o => o.status === 'fail').length;
  const skipN = outcomes.filter(o => o.status === 'skip').length;
  const errN  = outcomes.filter(o => o.status === 'error').length;
  console.log('VBP v1 conformance (Node skeleton)');
  console.log(`  tests:  ${outcomes.length}`);
  console.log(`  pass:   ${passN}`);
  console.log(`  fail:   ${failN}`);
  console.log(`  skip:   ${skipN}`);
  console.log(`  error:  ${errN}`);
  console.log(`  report: ${opts.out}`);
  process.exit((failN + errN) > 0 ? 1 : 0);
}

main();
